packages/server/src/loop.ts

```diff
@@ -34,160 +39,161 @@ export async function chargeCustomerInvoice({
+  if (isChargingSubscriptions) {
     return;
   }
-  isCharging = true;
+  isChargingSubscriptions = true;
 
   try {
     const now = new Date();
-
     let page = 0;
 
     // eslint-disable-next-line no-constant-condition
     while (true) {
-      // get draft invoice from past periods
-      const invoices = await database.invoices.find(
-        { date: { $lte: now }, status: 'draft' },
+      // get due subscriptions
+      const subscriptions = await database.subscriptions.find(
+        { nextPayment: { $lte: now }, status: 'active' },
         {
           limit: pageSize,
           offset: page * pageSize,
-          populate: ['project', 'items', 'subscription.changes', 'subscription.customer'],
+          populate: ['project', 'changes', 'customer'],
         },
       );
 
-      for await (let invoice of invoices) {
-        // Lock invoice processing
-        invoice.status = 'pending';
-        await database.em.persistAndFlush(invoice);
+      for await (const subscription of subscriptions) {
+        // TODO: should we lock subscription processing?
 
-        const { project, subscription } = invoice;
-        if (!subscription) {
-          log.error({ invoiceId: invoice._id }, 'Invoice has no subscription');
-          throw new Error('Invoice has no subscription');
-        }
+        const { project, customer } = subscription;
 
-        const { customer } = subscription;
+        const billingPeriod = getPeriodFromAnchorDate(subscription.nextPayment, subscription.anchorDate);
 
         try {
-          await database.em.populate(customer, ['activePaymentMethod']);
-          if (!customer.activePaymentMethod) {
-            log.error({ invoiceId: invoice._id, customerId: customer._id }, 'Customer has no active payment method');
-            throw new Error('Customer has no active payment method');
+          const existingInvoices = await database.invoices.find({
+            date: { $gte: billingPeriod.start, $lte: billingPeriod.end },
+            subscription,
+          });
+          if (existingInvoices.length > 0) {
+            log.error(
+              { subscriptionId: subscription._id, customerId: customer._id },
+              'Invoice for period already exists',
+            );
+            // TODO: should we just ignore this? currently this sets the subscription to error state?
+            throw new Error('Invoice for period already exists');
           }
 
-          invoice = addSubscriptionChangesToInvoice(subscription, invoice);
-          await chargeCustomerInvoice({ customer, invoice });
+          customer.invoiceCounter += 1;
+
+          const invoice = new Invoice({
+            currency: project.currency,
+            vatRate: project.vatRate,
+            sequentialId: customer.invoiceCounter,
+            subscription,
+            project,
+            status: 'draft',
+            date: billingPeriod.end, // TODO: or has this to be the current date when the invoice is issued?
```

Comment:
@anbraten the invoice date ust be the current date.

Reply:
Changed it. So worst case if the server is offline broken for 3 months. The invoice would be issued 3 months late, but for the old period (will be written that way in the invoice item)

Reply:
@anbraten Perfect